### ISSUE TITLE
Runtime V2 comes with missing and wrong assemblies

### ISSUE THREAD

**Issue by dersia** [Azure/Azure-Functions#625]
I just realized that running a V2 function ends up with bunch of errors.

the runtime is complaining about missing assembly ```Microsoft.AspNetCore.Mvc.Abstractions```
After checkoing %LOCALAPPDATA%\Azure.Functions.V2.Cli\2.0.1-beta I could confirm that the assembly was indeed missing.

After bringing in the assembly in version ```2.0.1``` it finally could find the assembly and was happy,
until it tried to load ```Microsoft.AspNetCore.Http.Abstraction```. Checkingfor that assembly I saw that it is there, but in version ```2.0.0``` and the runtime was looking for version ```2.0.1``` after fixing this, I had to do the same for all ```Microsoft.AspNetCore.Http.*``` packages.

After updating everything it finally went through fine.

**Comment by dersia** [comment]
after going on, I realized, that there is much more effected by this. All the AspNetCore and Mvc and EFCore etc packages need to be updated to version ```2.0.1```.

This is due to my update to ```Microsoft.NET.Sdk.Functions``` to version ```1.0.7``` which is the latest. It hase a dependency to ```Microsoft.Azure.WebJobs``` version ```3.0.0-beta4``` which has a dependency to versions ```2.0.1``` of all the AspNetCore packages.

Is code of the runtime anywhere available? I'd love to contribute to the runtime and do the update to the latest WebJobs version

**Comment by justcla** [comment]
Have you tried building and running the function locally? The tools, templates and CLI all need to be in sync. When you build and run the function locally, it should attempt to fetch all the new dependencies required.

**Comment by dersia** [comment]
Sorry for the confusion, I'm running this locally.

**Comment by justcla** [comment]
OK. I'll look into it. I'm on PST so you probably won't hear anything for another 12 hours.

**Comment by dersia** [comment]
Lol I'm pst too, so it's fine.

If I stay with 1.0.6 everything is fine. I'm wondering, if this is an issue when running once and then updating. I'll investigate too and provide more info tomorrow.

Good night!

**Comment by justcla** [comment]
Advice from engineering:

> The template that ships in VS still targets 1.0.6 version. 
> In this case the user has manually updated the version to 1.0.7 – 
>>This is due to my update to Microsoft.NET.Sdk.Functions to version 1.0.7

Sounds like you should stick with 1.0.6 :-)

**Comment by patricklee2** [comment]
The code for the runtime is available at https://github.com/Azure/azure-webjobs-sdk-script

**Comment by fabiocav** [comment]
Do you see this behavior if you run the runtime standalone (without the CLI)? This could be a CLI issue as the runtime itself keeps up with the SDK updates and dependencies. I'll take a look at your scenario, but the code is where @patricklee2 pointed you to and has the expected assembly references.

**Comment by ahmelsayed** [comment]
I think part of the problem is that VS doesn't update the runtime. `Microsoft.NET.Sdk.Functions` version `1.0.7` only works with `2.0.1-beta.22` cli, and I think vs pulls in a pretty old one. You can grab the beta.22 bits from here https://github.com/Azure/azure-functions-cli/releases/tag/2.0.1-beta.22 and replace the ones in the cli folder if you want to run on the latest v2 runtime.

**Comment by dersia** [comment]
I guess it's the cli. You can check this by installing the runtime, but not running the cli at all. When you check the V2 cli folder in %localappdata% you see that its empty. As soon as you hit debug (start he cli) you'll see that folder filling up with the runtime, but it is catching 1.0.6, even if your functions dependency is on 1.0.7. So it behaves the way mentioned by @justcla. 

I guess the cli needs to pull the runtime depending on the functions dependencies. This will probably end up with different runtime folders and the cli has to switch based on this, but I guess there is no better solution.

**Comment by dersia** [comment]
@ahmelsayed sounds good, I'll give it a try

**Comment by justcla** [comment]
The first thing to note is that v1.0.6 is the currently supported version of Functions SDK in Visual Studio.
And this is why the Functions extension in Visual Studio tries to pull down the 1.0.6 bits.

The next thing to note is that there is a strong dependency between the Functions SDK, the runtime, the CLI and the templates. If you change one, you must change them all to keep them in sync.
The CLI will bring down the corresponding runtime, so if you change Functions SDK version from 1.0.6 to 1.0.7, then just go fetch the corresponding CLI for 1.0.7 and at build time, it will fetch the correct runtime.
But you will also need to manually update the functions project (generated by the templates) to reference 1.0.7.

**Comment by justcla** [comment]
Currently, when building with Visual Studio, regardless of what Functions SDK version you've specified, it will run against a the version of the CLI that it wants. For V1 functions, this comes from a feed, which always has the latest CLI. For V2 functions, it is currently bound to the version that comes with the Functions extension.

Side note: You first get the Functions extension by including it in the Visual Studio installer (i.e. by choosing the Azure workload). This can later be updated by pulling down newer versions from [Marketplace](https://marketplace.visualstudio.com/items?itemName=VisualStudioWebandAzureTools.AzureFunctionsandWebJobsTools).)

Since V2 functions can only run in Visual Studio with the CLI shipped with the Visual Studio extension for Functions, and that is currently targeting Functions SDK 1.0.6, what that means is that Visual Studio will not be able to build against the CLI or runtimes associated with Functions SDK v1.0.7.

Note: This is not an issue if building from command line. Only an issue if building/running inside Visual Studio.

So, in short, if you want to build and run V2 Functions in Visual Studio, you must stick with Functions SDK v1.0.6 (for now).

**Comment by kashimiz** [comment]
I was able to force Visual Studio to use a Functions SDK v1.0.7-compatible version of the CLI using the following workaround:

1. Install the Azure Functions Core Tools 2.0 [per instructions here.
](https://docs.microsoft.com/en-us/azure/azure-functions/functions-run-local)
2. In VS, right-click project > Properties > Debug and set the following:

- **Launch:** Executable
- **Executable:** {full path of .NET Core CLI’s dotnet.exe, ex. C:\Program Files\dotnet\dotnet.exe}
- **Application Arguments:** {full path of the Azure Functions CLI assembly you want to run, ex. C:\Users\SomeUser\AppData\Roaming\npm\node_modules\azure-functions-core-tools\bin\Azure.Functions.Cli.dll} host start
- **Working Directory:** {full path of your project’s build output folder, ..\bin\Debug\netstandard2.0}

**Comment by wartmanrp** [comment]
@kashimiz This workaround is no longer working for me as of v1.0.9. When I try to launch with an exe debug profile I get the generic error:

> The program '[14116] dotnet.exe' has exited with code -2147450751 (0x80008081).

Any ideas on what could be causing this?

**Comment by SayusiAndo** [comment]
Similar here. I experience another issue which can be related to this.

My methods throws the error below. The strange is that the mentioned Microsoft.Extensions.Configuration.Abstractions is in the directory where other dlls are and the version number matches. I have been investigating it for hours without any clue why. One thing is sure it is related to config json file read, but this code worked a few days ago. If I set the code back to particular commits the issue do not disappear.
```
[3/20/2018 4:03:54 PM] Host lock lease acquired by instance ID '000000000000000000000000F559816F'.
[3/20/2018 4:03:55 PM] Job host started
[3/20/2018 4:04:41 PM] Function started (Id=738b492d-954f-4133-9f16-49bffd28e99d)
[3/20/2018 4:04:41 PM] Executing 'CountryV1_GetAllAsync_HttpStart' (Reason='This function was programmatically called via the host APIs.', Id=738b492d-954f-4133-9f16-49bffd28e99d)
[3/20/2018 4:04:41 PM] A ScriptHost error has occurred
[3/20/2018 4:04:41 PM] Exception while executing function: CountryV1_GetAllAsync_HttpStart. AzureDurableFunctions: Could not load file or assembly 'Microsoft.Extensions.Configuration.Abstractions, Version=2.0.1.0, Culture=neutral, PublicKeyToken=adb9793829ddae60'. Could not find or load a specific file. (Exception from HRESULT: 0x80131621). System.Private.CoreLib: Could not load file or assembly 'Microsoft.Extensions.Configuration.Abstractions, Version=2.0.1.0, Culture=neutral, PublicKeyToken=adb9793829ddae60'.
[3/20/2018 4:04:41 PM] Exception while executing function: CountryV1_GetAllAsync_HttpStart
[3/20/2018 4:04:41 PM] Exception while executing function: CountryV1_GetAllAsync_HttpStart. AzureDurableFunctions: Could not load file or assembly 'Microsoft.Extensions.Configuration.Abstractions, Version=2.0.1.0, Culture=neutral, PublicKeyToken=adb9793829ddae60'. Could not find or load a specific file. (Exception from HRESULT: 0x80131621). System.Private.CoreLib: Could not load file or assembly 'Microsoft.Extensions.Configuration.Abstractions, Version=2.0.1.0, Culture=neutral, PublicKeyToken=adb9793829ddae60'.
[3/20/2018 4:04:41 PM] Function completed (Failure, Id=738b492d-954f-4133-9f16-49bffd28e99d, Duration=168ms)
[3/20/2018 4:04:41 PM]
[3/20/2018 4:04:41 PM] Executed 'CountryV1_GetAllAsync_HttpStart' (Failed, Id=738b492d-954f-4133-9f16-49bffd28e99d)
[3/20/2018 4:04:41 PM] System.Private.CoreLib: Exception while executing function: CountryV1_GetAllAsync_HttpStart. AzureDurableFunctions: Could not load file or assembly 'Microsoft.Extensions.Configuration.Abstractions, Version=2.0.1.0, Culture=neutral, PublicKeyToken=adb9793829ddae60'. Could not find or load a specific file. (Exception from HRESULT: 0x80131621). System.Private.CoreLib: Could not load file or assembly 'Microsoft.Extensions.Configuration.Abstractions, Version=2.0.1.0, Culture=neutral, PublicKeyToken=adb9793829ddae60'.
[3/20/2018 4:04:41 PM]   Function had errors. See Azure WebJobs SDK dashboard for details. Instance ID is '738b492d-954f-4133-9f16-49bffd28e99d'
[3/20/2018 4:04:41 PM] System.Private.CoreLib: Exception while executing function: CountryV1_GetAllAsync_HttpStart. AzureDurableFunctions: Could not load file or assembly 'Microsoft.Extensions.Configuration.Abstractions, Version=2.0.1.0, Culture=neutral, PublicKeyToken=adb9793829ddae60'. Could not find or load a specific file. (Exception from HRESULT: 0x80131621). System.Private.CoreLib: Could not load file or assembly 'Microsoft.Extensions.Configuration.Abstractions, Version=2.0.1.0, Culture=neutral, PublicKeyToken=adb9793829ddae60'.
```

**Comment by wartmanrp** [comment]
@SayusiAndo I get a similar error (except about ms.extensions.mvc.abstractions) even though all of the assemblies are the correct version. I can get the project to build if I use the project launch debug profile, but with the error, if I try executable launch as detailed in the workaround, I get the generic error I mentioned.

It is worth noting, as pointed out earlier in this thread, if I build a new project from the functions CLI everything works just great, but I then just have to use VS Code instead of 2017CE, which is fine by me.

**Comment by cpwood** [comment]
Struggling with this today too.

I've managed to get debug and breakpoints working with 1.0.9 in Visual Studio using these settings:

**Launch:** Executable
**Executable:** %APPDATA%\npm\node_modules\azure-functions-core-tools\bin\func.exe
**Application Arguments:** host start
**Working Directory:** $(TargetDir)

But I can't get my function to work because of conflicting DLL versions between the functions runtime and my referenced .NET Standard libraries (relating to Microsoft.Extensions.Configuration.dll v2.0.1).

**Comment by TyrinCRM** [comment]
Also struggling with this. I've tried all the variations above for override launch settings.

**Comment by patricklee2** [comment]
@justcla do you have an update?

**Comment by nzthiago** [comment]
While the Visual Studio extension doesn't get updated to fix this, you need to configure the VS Debug settings to point to func.dll instead of Azure.Functions.Cli.dll. I have it working with the following Debug settings for my Functions project:

**Launch:** Executable
**Executable:** C:\Program Files\dotnet\dotnet.exe  _(or wherever you have dotnet installed)_
**Application Arguments:** %APPDATA%\Roaming\npm\node_modules\azure-functions-core-tools\bin\func.dll host start
**Working Directory:** $(TargetDir)

Note, not sure if this matters but I have:
* Latest core tools (to version 2.0.1-beta.25)
* Updated the SDK and other packages as per the [latest announcement](https://github.com/Azure/app-service-announcements/issues/94#issuecomment-376647094) and using that from my function

**Comment by danroot** [comment]
I have this same issue, but the workaround listed by @nzthiago and others does not resolve it for me.  I have gotten the latest azure functions bits, ensured the project is "v2", and changed the debug to run dotnet %APPDATA%\npm\node_modules\azure-functions-core-tools\bin\func.dll host start.  I can run a simple function, however when I reference a custom .netstandard2.0 library project that in turn references Microsoft.AspNetCore.Mvc 2.0.2 or 2.0.3, and run, I get the same error as above.  Adding NuGet packages for the "missing" assemblies and rebuilding the app does not resolve the issue.  Removing the custom .netstandard2.0 library reference "fixes" the issue, but then I can't reuse the model classes, etc. I need in the function app.

**Comment by TravisLRiffle** [comment]
In short, I'm having the same issue.

**Comment by SoulFireMage** [comment]
I've the same issue as well. Can't find a sensible way round - I'm only running an example from Samples library.

![image](https://user-images.githubusercontent.com/2396080/39365030-e7265b86-4a27-11e8-8209-87a88abb3c4c.png)

ASP.NET Core packages): System.Private.CoreLib: Could not load file or assembly 'Microsoft.AspNetCore.Hosting.Server.Abstractions, Version=2.0.30 #2364

Not really sure what else I can try with this?

**Comment by chmarti** [comment]
So many good things coming in Functions v2, but right now this issue is blocking me from using them as well.

**Comment by chmarti** [comment]
I finally got it working by by going the same route as @vsadams. Luckily it isn't the end of the world for me to use the downgraded libraries, but this is still annoying.

**Comment by gabrieljoelc** [comment]
Per @vsadams and @chmarti 's suggestions, I downgraded `Microsoft.NET.Sdk.Functions` from 1.0.13 to 1.0.6 and stopped getting this error.

**Comment by neoKushan** [comment]
I had this same issue, however I upgraded the "Azure Functions and Web Jobs Tools"  extension from 15.0.40215.0 to 15.0.40502.2 and the issue went away while still using Microsoft.NET.Sdk.Functions v1.0.13 (latest at time of posting).

Also running Visual Studio 15.6.3, for what it's worth.

**Comment by vtjc2002** [comment]
@neoKushan 's solution works for me as wel

**Comment by johnwc** [comment]
I have 15.0.40502.**0** installed and I am still seeing this issue, where did you find 15.0.40502.**2**?

**Comment by johnwc** [comment]
Is this issue related to this one? https://github.com/Azure/azure-functions-core-tools/issues/335

**Comment by MisinformedDNA** [comment]
I'm having the same issue. How do I upgrade to 15.0.40502.2?

**Comment by neoKushan** [comment]
The correct version is **15.0.40502.0**, not 15.0.40502.2 as I had originally typed. My apologies.

**Comment by mishal153** [comment]
I found the same issue when trying on VS code(unable to fix) and visual studio(fixed).
On VS code I have tried updating the [azure core tools](https://github.com/Azure/azure-functions-core-tools) to version  2.0.1-beta1.15 but issue remains.
Any suggestions ?

Visual studio: I was able to solve it on Visual studio by updating the 'Azure functions tools and webjobs tools' to 15.0.40502.0 as mentioned above.

**Comment by paulbatum** [comment]
@mishal153 It looks like your copy of core tools is well out of date. See [this releases page](https://github.com/Azure/azure-functions-core-tools/releases) - the latest at time of writing is beta1.26

**Comment by mishal153** [comment]
You are right. Thanks. That solved it. I was kind of guided by the 'wanted' column of the `npm outdated` command. Thanks

 
![capture](https://user-images.githubusercontent.com/6596643/40591446-cedf719e-6254-11e8-8b2c-f5c4e9a9f4e8.PNG)

**Comment by PankajPatro** [comment]
Using VS2017. Create a Sample .NetCore Azure Function App. Launch and see it as working. Now go to the Nuget Package Manger and see if there are any updates for nuget packages. Update the Azure SDK to the latest and voila, the Azure Functions stops working with the same issue Microsoft.AspNetCore.Mvc.Abstractions

**Comment by paulbatum** [comment]
@PankajPatro thats a bug, tracked here https://github.com/Azure/azure-webjobs-sdk/issues/1942
